src/proper_sa.erl

```diff
@@ -28,80 +28,178 @@
-         temp_func = fun(_, _, _, _, _) -> 1.0 end   :: temp_fun(),
-         %% output function
-         output_fun = fun (_, _) -> ok end            :: output_fun()}).
+        {%% search steps
+         k_max = 0                                 :: k(),
+         %% current step
+         k_current = 0                             :: k(),
+         %% acceptance function
+         p = fun (_, _, _) -> false end            :: accept_fun(),
+         %% fitness
+         last_energy = null                        :: proper_target:fitness() | null,
+         last_update = 0                           :: integer(),
+         %% temperature
+         temperature = 1.0                         :: proper_gen_next:temperature(),
+         temp_func = fun(_, _, _, _, _) -> 1.0 end :: temp_fun()}).
+-type sa_data() :: #sa_data{}.
+
+%% -----------------------------------------------------------------------------
+%% proper_target callbacks
+%% -----------------------------------------------------------------------------
+
+%% Initialize the strategy data based on the 
+%% number of the search steps and the strategy.
+
+%% @private
+-spec init_strategy(proper:setup_opts()) -> sa_data().
+init_strategy(#{numtests := Steps}) ->
+  #sa_data{k_max = Steps,
+           p = get_acceptance_function(),
+           temp_func = get_temperature_function()}.
+
+%% Initialize target state based on the initial generator
+%% and the neighbourhood function.
+
+%% @private
+-spec init_target(proper_target:tmap()) -> sa_target().
+init_target(#{first := First, next := Next}) ->
+  {ok, InitialValue} = proper_gen:safe_generate(First),
+  #sa_target{first = First, next = Next, last_generated = InitialValue}.
+
+%% The function which generates the next instances of
+%% the targeted generator. It also updates the target state.
+
+%% @private
+-spec next(sa_target(), sa_data()) -> {any(), sa_target(), sa_data()}.
+next(Target, Data) ->
+  Temperature = Data#sa_data.temperature,
+  NextGenerator = (Target#sa_target.next)(Target#sa_target.last_generated,
+                                          Temperature),
+  {ok, Generated} = proper_gen:safe_generate(NextGenerator),
+  {Generated, Target#sa_target{current_generated = Generated}, Data}.
+
+%% Update state and data based on current fitness.
+%% The current generated value is accepted based on the
+%% simulated annealing acceptance function, which always
+%% accepts better fitnesses, while accepting worst fitnesses
+%% based on the acceptance probability.
+
+%% @private
+-spec update_fitness(proper_target:fitness(), sa_target(), sa_data()) ->
+  {sa_target(), sa_data()}.
+update_fitness(Fitness, Target, Data) ->
+  #sa_data{k_current = K_Current,
+           k_max = K_Max,
+           last_energy = Energy,
+           temperature = Temperature,
+           temp_func = TempFunc} = Data,
+  case (Energy =:= null)
+       orelse (Data#sa_data.p)(Energy,
```

Comment:
Why isn't the `Data#sa_data.p` field taken out as part of the record deconstruction above?
if you change that, then most likely the whole case will fit in one line, instead of being across 4 of them!